### ISSUE TITLE
[Feature Request] Grant all permissions to test apk too

### ISSUE THREAD

**Issue by tir38** [square/spoon#537]
What:
The `-grant-all` flag grants all permissions to apk under test. It would be great if this flag granted all permissions to the *test* apk as well.

Reason:
Sometimes the test apk needs permissions that the apk under test does not. This is true especially with Spoon which needs `android.permission.WRITE_EXTERNAL_STORAGE`. If the app itself doesn't need this permission, then the right thing to do is put that permission in the `androidTest/AndroidManifest.xml`. However there is no way to grant this permission during test.

A more detailed explanation of the problem is [here](https://stackoverflow.com/q/49791104)

Solution:
If I understand `SpoonDeviceRunner` correctly, we should be able to just duplicate the `-g` addition for the test apk:

https://github.com/square/spoon/blob/1e6674f3a0c4d896b68ac789d4f71a6a09646d0f/spoon-runner/src/main/java/com/squareup/spoon/SpoonDeviceRunner.java#L161

```
    // Now install the main application and the instrumentation application.
    for (File otherApk : otherApks) {
      try {
        String extraArgument = "";
        if (grantAll && deviceDetails.getApiLevel() >= DeviceDetails.MARSHMALLOW_API_LEVEL) {
          extraArgument = "-g";
        }
        device.installPackage(otherApk.getAbsolutePath(), true, extraArgument);
      } catch (InstallException e) {
        logInfo("InstallException while install other apk on device [%s]", serial);
        e.printStackTrace(System.out);
        return result.markInstallAsFailed("Unable to install other APK.").addException(e).build();
      }
    }
    try {
        String extraArgument = "";
        if (grantAll && deviceDetails.getApiLevel() >= DeviceDetails.MARSHMALLOW_API_LEVEL) {
          extraArgument = "-g";
        }
        device.installPackage(testApk.getAbsolutePath(), true, extraArgument);
    } catch (InstallException e) {
      logInfo("InstallException while install test apk on device [%s]", serial);
      e.printStackTrace(System.out);
      return result.markInstallAsFailed("Unable to install instrumentation APK.")
          .addException(e)
          .build();
    }
```

**Comment by edenman** [comment]
Sounds reasonable.  Make a PR?

**Comment by tir38** [comment]
will do.

**Comment by tir38** [comment]
Looking at the code my initial changes are to extract

```
    /**
     * @param deviceDetails details for current device
     * @return optional argument for grantAll permissions, or empty string if not required
     */
    private String getGrantAllExtraArgument(DeviceDetails deviceDetails) {
        String extraArgument = "";
        if (grantAll && deviceDetails.getApiLevel() >= DeviceDetails.MARSHMALLOW_API_LEVEL) {
            extraArgument = "-g";
        }
        return extraArgument;
    }
```

and then call it twice:

      // Now install the main application and the instrumentation application.
        for (File otherApk : otherApks) {
            try {
                String extraArgument = getGrantAllExtraArgument(deviceDetails);
                device.installPackage(otherApk.getAbsolutePath(), true, extraArgument);
                 ...
        }
        try {
            String extraArgument = getGrantAllExtraArgument(deviceDetails);
            device.installPackage(testApk.getAbsolutePath(), true, extraArgument);
            ...
        }

This solves the general problem of granting all permissions to the test apk. Then I saw this method

```
private void grantReadWriteExternalStorage(DeviceDetails deviceDetails, IDevice device)
```

which gets called automatically (regardless of -grantAll). Would I be better off updating this method to also grand read/write permission on test apk? This would solve my *specific* problem.

**Comment by edenman** [comment]
I'd lean toward just applying the `-g` flag to both

**Comment by edenman** [comment]
Thanks!

**Comment by tir38** [comment]
I assume this will go out with the next `SNAPSHOT`?

**Comment by edenman** [comment]
yup.  should be automatically created if it's not already

**Comment by tir38** [comment]
I guess this is a separate issue. I'll open another ticket if I need to.

**Comment by tir38** [comment]
TL;DR `grantAll` for the test apk surely has some benefit, but not if all you want to do is move the `WRITE_STORAGE_PERMISSION` into your test apk's manifest (https://github.com/square/spoon/issues/537)
